lib/elixir/src/elixir_expand.erl

```diff
@@ -561,12 +561,19 @@ var_context(Meta, Kind) ->
 expand_case(true, Meta, Expr, Opts, E) ->
   {EExpr, EE} = expand(Expr, E),
   {EOpts, EO} = elixir_clauses:'case'(Meta, Opts, EE),
+
   ROpts =
-    case proplists:get_value(optimize_boolean, Meta, false) andalso
-         elixir_utils:returns_boolean(EExpr) of
-      true -> rewrite_case_clauses(EOpts);
-      false -> EOpts
+    case proplists:get_value(optimize_boolean, Meta, false) of
+      true ->
+        case elixir_utils:returns_boolean(EExpr) of
+          true -> rewrite_case_clauses(EOpts);
+          false -> generated_case_clauses(EOpts)
```

Comment:
I don't know if it's better to introduce another meta like `generated_clauses`, but this is necessary to not raise warnings and `optimize_boolean` is used only on `Kernel`